File: sdks/php/src/Api/Topics.php
<?php

declare(strict_types=1);

namespace Apollo\Deploy\Signal\Sdk\Api;

use Apollo\Deploy\Signal\Sdk\SDKError;
use Apollo\Deploy\Signal\Sdk\SSEEvent;
use Apollo\Deploy\Signal\Sdk\Transport;
use Apollo\Deploy\Signal\Sdk\CreateTopicBody;
use Apollo\Deploy\Signal\Sdk\CursorPage;
use Apollo\Deploy\Signal\Sdk\TopicPageResponse;
use Apollo\Deploy\Signal\Sdk\TopicResponse;
use Apollo\Deploy\Signal\Sdk\UpdateTopicBody;
use Psr\Http\Message\ResponseInterface;

/** TopicsAPI domain operations. */
class Topics
{
    public function __construct(private Transport $transport)
    {
    }

    /**
     * @param string $projectId
     * @return TopicPageResponse
     * @throws SDKError
     */
    public function listTopics(
        string $projectId,
    ): TopicPageResponse
    {
        $path = '/v1/projects/${projectId}/topics';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(TopicPageResponse::class, 'fromArray')) {
            return TopicPageResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $topicId
     * @return TopicResponse
     * @throws SDKError
     */
    public function getTopic(
        string $projectId,
        string $topicId,
    ): TopicResponse
    {
        $path = '/v1/projects/${projectId}/topics/${topicId}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${topicId}', rawurlencode(Transport::wireValue($topicId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(TopicResponse::class, 'fromArray')) {
            return TopicResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $topicId
     * @return CursorPage
     * @throws SDKError
     */
    public function listContactsInTopic(
        string $projectId,
        string $topicId,
    ): CursorPage
    {
        $path = '/v1/projects/${projectId}/topics/${topicId}/contacts';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${topicId}', rawurlencode(Transport::wireValue($topicId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(CursorPage::class, 'fromArray')) {
            return CursorPage::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param CreateTopicBody $input
     * @return TopicResponse
     * @throws SDKError
     */
    public function createTopic(
        string $projectId,
        CreateTopicBody $input,
    ): TopicResponse
    {
        $path = '/v1/projects/${projectId}/topics';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'POST',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];
        if ($input !== null) {
            $request['body'] = $input;
            $request['contentType'] = 'application/json';
        }

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(TopicResponse::class, 'fromArray')) {
            return TopicResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $topicId
     * @param UpdateTopicBody $input
     * @return TopicResponse
     * @throws SDKError
     */
    public function updateTopic(
        string $projectId,
        string $topicId,
        UpdateTopicBody $input,
    ): TopicResponse
    {
        $path = '/v1/projects/${projectId}/topics/${topicId}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${topicId}', rawurlencode(Transport::wireValue($topicId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'PATCH',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];
        if ($input !== null) {
            $request['body'] = $input;
            $request['contentType'] = 'application/json';
        }

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(TopicResponse::class, 'fromArray')) {
            return TopicResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $topicId
     * @return void
     * @throws SDKError
     */
    public function deleteTopic(
        string $projectId,
        string $topicId,
    ): void
    {
        $path = '/v1/projects/${projectId}/topics/${topicId}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${topicId}', rawurlencode(Transport::wireValue($topicId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'DELETE',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $this->transport->readResponseBody($response);
        return;
    }

    private function decodeResponse(ResponseInterface $response): mixed
    {
        $contents = $this->transport->readResponseBody($response);
        if (trim($contents) === '') {
            $requestId = $response->getHeaderLine('x-request-id')
                ?: $response->getHeaderLine('request-id')
                ?: null;
            throw new SDKError(
                message: 'Expected a response body for a non-void operation',
                status: $response->getStatusCode(),
                code: 'RESPONSE_BODY_MISSING',
                requestId: $requestId,
            );
        }

        try {
            return json_decode($contents, true, 512, JSON_THROW_ON_ERROR);
        } catch (\JsonException) {
            return $contents;
        }
    }
}

File: sdks/rust/src/api/topics.rs
use crate::error::SdkError;
use crate::transport::Transport;
use reqwest::Method;
use serde::Serialize;
use crate::types::{ CreateTopicBody, CursorPage, TopicPageResponse, TopicResponse, UpdateTopicBody };

/// TopicsAPI API operations.
pub struct TopicsAPI {
    transport: Transport,
}

impl TopicsAPI {
    pub(crate) fn new(transport: Transport) -> Self {
        Self { transport }
    }

    pub async fn list_topics(
        &self,
        project_id: String,
    ) -> Result<TopicPageResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/topics".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<TopicPageResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn get_topic(
        &self,
        project_id: String,
        topic_id: String,
    ) -> Result<TopicResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/topics/${topicId}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${topicId}", &encode_path(&topic_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<TopicResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn list_contacts_in_topic(
        &self,
        project_id: String,
        topic_id: String,
    ) -> Result<CursorPage, SdkError> {
        let mut path = "/v1/projects/${projectId}/topics/${topicId}/contacts".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${topicId}", &encode_path(&topic_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<CursorPage>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn create_topic(
        &self,
        project_id: String,
        body: &CreateTopicBody,
    ) -> Result<TopicResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/topics".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<TopicResponse>(
            Method::POST,
            &path,
            query,
            Some(body),
            headers,
            None,
        ).await
    }

    pub async fn update_topic(
        &self,
        project_id: String,
        topic_id: String,
        body: &UpdateTopicBody,
    ) -> Result<TopicResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/topics/${topicId}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${topicId}", &encode_path(&topic_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<TopicResponse>(
            Method::PATCH,
            &path,
            query,
            Some(body),
            headers,
            None,
        ).await
    }

    pub async fn delete_topic(
        &self,
        project_id: String,
        topic_id: String,
    ) -> Result<(), SdkError> {
        let mut path = "/v1/projects/${projectId}/topics/${topicId}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${topicId}", &encode_path(&topic_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute_raw(
            Method::DELETE,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await.map(|_| ())
    }

}

fn encode_path(value: &str) -> String {
    urlencoding::encode(value).into_owned()
}

fn encode_value<T: Serialize>(value: &T) -> Result<String, SdkError> {
    match serde_json::to_value(value)
        .map_err(|error| SdkError::ValidationError(error.to_string()))?
    {
        serde_json::Value::String(value) => Ok(value),
        value => Ok(value.to_string()),
    }
}

fn encode_cookie_name(name: &str) -> String {
    urlencoding::encode(name).into_owned()
}

fn encode_cookie_value<T: Serialize>(value: &T) -> Result<String, SdkError> {
    let value = encode_value(value)?;
    Ok(urlencoding::encode(&value).into_owned())
}
